### PR TITLE
Allow writing request types in handler signatures

### PR DESCRIPTION
**Pull request by neongreen**
Figuring out "which request body type does this handler take" is a constant annoyance for me. Should be easier with this change.

If you think it's worth it, I'll migrate the rest.

### CODE REVIEW

**Comment by neongreen** [comment]
Before and after:

```diff
      post "/onboarding/v3" (continue onboarding) $
          accept "application" "json"
-         .&. contentType "application" "json"
          .&. header "Z-User"
-         .&. request
+         .&. jsonRequest @AddressBook
```

```diff
- onboarding :: JSON ::: JSON ::: UserId ::: Request -> Handler Response
- onboarding (_ ::: _ ::: uid ::: r) = do
+ onboarding :: JSON ::: UserId ::: JsonRequest AddressBook -> Handler Response
+ onboarding (_ ::: uid ::: r) = do
      ab <- parseJsonBody r
      json <$> API.onboarding uid ab !>> connError
```

**Comment by neongreen** [comment]
And yeah, this would've been served better by Servant, but Servant is not coming any time soon.

File: services/brig/src/Brig/API/Handler.hs [hunk 1]
```diff
@@ -76,7 +76,7 @@ type JSON = Media "application" "json"
 -- TODO: move to libs/wai-utilities?  there is a parseJson' in "Network.Wai.Utilities.Request",
 -- but adjusting its signature to this here would require to move more code out of brig (at least
 -- badRequest and probably all the other errors).
-parseJsonBody :: FromJSON a => Request -> Handler a
+parseJsonBody :: FromJSON a => JsonRequest a -> Handler a
```
Comment (ChrisPenner) [inline services/brig/src/Brig/API/Handler.hs]: Nice! I ran into a copy paste error the other day that this would have prevented 👍 MORE TYPES! Yay!

**Comment by ChrisPenner** [comment]
Looks like a net win to me; not really seeing any downsides here. Good stuff! One step closer to auto-generating those pesky swagger docks mwahaha....

**Comment by jschaul** [comment]
> If you think it's worth it, I'll migrate the rest.

I think this is a big improvement and it'd be worth migrating the rest.

**Pull request merged**